src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfJsonPrimitive.cs

```diff
@@ -0,0 +1,270 @@
+                    : (rented = ArrayPool<byte>.Shared.Rent(quotedValueLength));
+                buffer = buffer.Slice(0, quotedValueLength);
+
+                buffer[0] = buffer[buffer.Length - 1] = JsonConstants.Quote;
+                _value.CopyTo(buffer.Slice(1, _value.Length));
+
+                writer.WriteRawValue(buffer);
+            }
+            finally
+            {
+                if (rented is not null)
+                {
+                    ArrayPool<byte>.Shared.Return(rented);
+                }
+            }
+        }
+
+        public override T GetValue<T>()
+        {
+            if (!TryGetValue(out T? value))
+            {
+                ThrowHelper.ThrowInvalidOperationException_NodeUnableToConvertElement(_valueKind, typeof(T));
+            }
+
+            return value;
+        }
+
+        public override bool TryGetValue<T>([NotNullWhen(true)] out T? value) where T : default
+        {
+            bool success;
+
+            if (typeof(T) == typeof(JsonElement))
+            {
+                success = TryGetJsonElement(this, out JsonElement element);
+                value = (T)(object)element;
+                return success;
+            }
+
+            switch (_valueKind)
+            {
+                case JsonValueKind.Number:
+                    if (typeof(T) == typeof(int) || typeof(T) == typeof(int?))
+                    {
+                        success = Utf8Parser.TryParse(_value, out int result, out int bytesConsumed) &&
+                                  _value.Length == bytesConsumed;
+                        value = (T)(object)result;
+                        return success;
+                    }
+
+                    if (typeof(T) == typeof(long) || typeof(T) == typeof(long?))
+                    {
+                        success = Utf8Parser.TryParse(_value, out long result, out int bytesConsumed) &&
+                                  _value.Length == bytesConsumed;
+                        value = (T)(object)result;
+                        return success;
+                    }
+
+                    if (typeof(T) == typeof(double) || typeof(T) == typeof(double?))
+                    {
+                        success = Utf8Parser.TryParse(_value, out double result, out int bytesConsumed) &&
+                                  _value.Length == bytesConsumed;
+                        value = (T)(object)result;
+                        return success;
+                    }
+
+                    if (typeof(T) == typeof(short) || typeof(T) == typeof(short?))
+                    {
+                        success = Utf8Parser.TryParse(_value, out short result, out int bytesConsumed) &&
+                                  _value.Length == bytesConsumed;
+                        value = (T)(object)result;
+                        return success;
+                    }
+
+                    if (typeof(T) == typeof(decimal) || typeof(T) == typeof(decimal?))
```

Comment:
There are limits to how many types we can support with this branching. Assuming the `JsonNode` encapsulates a `JsonSerializerOptions` (I don't recall if it does, or can be made to do that) it might be possible to extend conversion support to arbitrary types with simpler logic here.